Compilation error in bpf_lxc when running on top of Flannel

Hi we are using hubble and was pointed to the cilium repository for this issue as cilium/hubble is used for the client-side CLI binary. We are using the hubble release version 1.9.4. and are trying to get setup with Hubble on our custom K8S cluster, but our Coredns pods are failing to start and are stuck in ContainerCreating

OS: Ubuntu 20, Kernel: 5.4.0-65-generic

`coredns-f9fd979d6-gxxd7 0/1 ContainerCreating
`

Describing the coredns pod provides this:

```
  ----     ------                  ----                ----               -------
  Normal   Scheduled               114s                default-scheduler  Successfully assigned kube-system/coredns-f9fd979d6-z7v2k to fh1-kubet04
  Warning  FailedCreatePodSandBox  23s                 kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "99219f93c41e772d90548436b45da1201c9d0804763b6267b9ba91bd6d8a7e76" network for pod "coredns-f9fd979d6-z7v2k": networkPlugin cni failed to set up pod "coredns-f9fd979d6-z7v2k_kube-system" network: unable to create endpoint: Cilium API client timeout exceeded
  Warning  DNSConfigForming        22s (x2 over 114s)  kubelet            Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is: 127.0.0.1 172.21.0.5 172.21.0.6
  Normal   SandboxChanged          22s                 kubelet            Pod sandbox changed, it will be killed and re-created.

```
It is complaining about: networkPlugin cni failed to set up pod "coredns-f9fd979d6-z7v2k_kube-system" network: unable to create endpoint: Cilium API client timeout exceeded

Cilium pod logs look like the following:

```
level=warning msg="Regeneration of endpoint failed" bpfCompilation=145.146053ms bpfLoadProg=0s bpfWaitForELF=145.427856ms bpfWriteELF=0s containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3217 error="Could not locate previously compiled BPF template" identity=50897 ipv4= ipv6= k8sPodName=/ mapSync="3.1µs" policyCalculation="4.8µs" prepareBuild="450.004µs" proxyConfiguration="8.6µs" proxyPolicyCalculation="20.201µs" proxyWaitForAck=0s reason="retrying regeneration" subsys=endpoint total=146.887271ms waitingForCTClean="592.905µs" waitingForLock="1.3µs"
level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3217 error="Could not locate previously compiled BPF template" identity=50897 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=error msg="Failed to compile bpf_lxc.o: exit status 1" compiler-pid=498 linker-pid=499 subsys=datapath-loader
level=warning msg="In file included from /var/lib/cilium/bpf/bpf_lxc.c:28:" subsys=datapath-loader
level=warning msg="In file included from /var/lib/cilium/bpf/lib/lxc.h:16:" subsys=datapath-loader
level=warning msg="/var/lib/cilium/bpf/lib/proxy.h:195:6: error: unused variable 'result' [-Werror,-Wunused-variable]" subsys=datapath-loader
level=warning msg="        int result = CTX_ACT_OK;" subsys=datapath-loader
level=warning msg="            ^" subsys=datapath-loader
level=warning msg="1 error generated." subsys=datapath-loader
level=warning msg="JoinEP: Failed to compile" debug=false error="Failed to compile bpf_lxc.o: exit status 1" params="&{Source:bpf_lxc.c Output:bpf_lxc.o OutputType:obj}" subsys=datapath-loader
level=error msg="BPF template object creation failed" bpfHeaderfileHash=2bc4c3266208be496ce852ce5ec4e33c41ae5f6c error="failed to compile template program /var/run/cilium/state/templates/2bc4c3266208be496ce852ce5ec4e33c41ae5f6c: Failed to compile bpf_lxc.o: exit status 1" subsys=datapath-loader
level=warning msg="BPF template compilation unsuccessful" bpfHeaderfileHash=2bc4c3266208be496ce852ce5ec4e33c41ae5f6c error="Could not locate previously compiled BPF template" subsys=datapath-loader
level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3217 error="Could not locate previously compiled BPF template" identity=50897 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=warning msg="generating BPF for endpoint failed, keeping stale directory." containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3217 file-path=3217_next_fail identity=50897 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=warning msg="Regeneration of endpoint failed" bpfCompilation=195.180454ms bpfLoadProg=0s bpfWaitForELF=195.330355ms bpfWriteELF=0s containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3217 error="Could not locate previously compiled BPF template" identity=50897 ipv4= ipv6= k8sPodName=/ mapSync="3.8µs" policyCalculation="6.401µs" prepareBuild="590.005µs" proxyConfiguration="19.8µs" proxyPolicyCalculation="26.7µs" proxyWaitForAck=0s reason="retrying regeneration" subsys=endpoint total=196.85327ms waitingForCTClean="600.906µs" waitingForLock="1.6µs"
level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3217 error="Could not locate previously compiled BPF template" identity=50897 ipv4= ipv6= k8sPodName=/ subsys=endpoint

```

Running cilium --verbose:

```
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.19 (v1.19.6) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Probe   [eth0 (Direct Routing)]
Cilium:                 Ok      OK
Controller Status:      12/14 healthy
  Name                                  Last success   Last error   Count   Message
  dns-garbage-collector-job             30s ago        never        0       no error
  endpoint-1593-regeneration-recovery   never          42s ago      28      regeneration recovery failed
  endpoint-46-regeneration-recovery     never          6s ago       7       regeneration recovery failed
  k8s-heartbeat                         31s ago        never        0       no error
  mark-k8s-node-as-available            13m23s ago     never        0       no error
  metricsmap-bpf-prom-sync              6s ago         never        0       no error
  resolve-identity-1593                 3m23s ago      never        0       no error
  resolve-identity-46                   50s ago        never        0       no error
  sync-endpoints-and-host-ips           23s ago        never        0       no error
  sync-lb-maps-with-k8s-services        13m23s ago     never        0       no error
  sync-to-k8s-ciliumendpoint (1593)     3s ago         never        0       no error
  sync-to-k8s-ciliumendpoint (46)       9s ago         never        0       no error
  template-dir-watcher                  never          never        0       no error
  update-k8s-node-annotations           13m29s ago     never        0       no error
Hubble:         Ok   Current/Max Flows: 0/4096 (0.00%), Flows/s: 0.00   Metrics: Disabled
KubeProxyReplacement Details:
  Status:              Probe
  Protocols:           TCP, UDP
  Devices:             eth0 (Direct Routing)
  Mode:                SNAT
  Backend Selection:   Random
  Session Affinity:    Enabled
  XDP Acceleration:    Disabled
  Services:
  - ClusterIP:      Enabled
  - NodePort:       Enabled
  - LoadBalancer:   Enabled
  - externalIPs:    Enabled
  - HostPort:       Enabled
Cluster health:         Warning   cilium-health daemon unreachable

```

Configmap is as follows:

```
apiVersion: v1
data:
  auto-direct-node-routes: "false"
  bpf-lb-map-max: "65536"
  bpf-map-dynamic-size-ratio: "0.0025"
  bpf-policy-map-max: "16384"
  cilium-endpoint-gc-interval: 5m0s
  cluster-id: ""
  cluster-name: default
  cluster-pool-ipv4-cidr: 10.0.0.0/8
  cluster-pool-ipv4-mask-size: "24"
  custom-cni-conf: "false"
  debug: "false"
  disable-cnp-status-updates: "true"
  enable-auto-protect-node-port-range: "true"
  enable-bandwidth-manager: "false"
  enable-bpf-clock-probe: "true"
  enable-bpf-masquerade: "true"
  enable-endpoint-health-checking: "true"
  enable-endpoint-routes: "true"
  enable-health-check-nodeport: "true"
  enable-health-checking: "true"
  enable-hubble: "true"
  enable-ipv4: "true"
  enable-ipv6: "false"
  enable-l7-proxy: "true"
  enable-local-redirect-policy: "false"
  enable-policy: default
  enable-remote-node-identity: "true"
  enable-session-affinity: "true"
  enable-well-known-identities: "false"
  enable-xt-socket-fallback: "true"
  flannel-master-device: cni0
  flannel-uninstall-on-exit: "false"
  hubble-socket-path: /var/run/cilium/hubble.sock
  identity-allocation-mode: crd
  install-iptables-rules: "true"
  ipam: cluster-pool
  kube-proxy-replacement: probe
  kube-proxy-replacement-healthz-bind-address: ""
  masquerade: "true"
  monitor-aggregation: medium
  monitor-aggregation-flags: all
  monitor-aggregation-interval: 5s
  native-routing-cidr: 10.244.0.0/16
  node-port-bind-protection: "true"
  operator-api-serve-addr: 127.0.0.1:9234
  preallocate-bpf-maps: "false"
  sidecar-istio-proxy-image: cilium/istio_proxy
  tunnel: disabled
  wait-bpf-mount: "false"

```
Any help/pointers would be appreciated!

--- a/bpf/Makefile
+++ b/bpf/Makefile
@@ -207,6 +207,7 @@ bpf_xdp.o: bpf_xdp.ll
 
 # The following option combinations are compile tested
 LXC_OPTIONS = \
+	 -DHOST_REDIRECT_TO_INGRESS: \
 	 -DALLOW_ICMP_FRAG_NEEDED: \
 	 -DSKIP_DEBUG: \
 	 -DHAVE_LPM_TRIE_MAP_TYPE: \

--- a/bpf/lib/proxy.h
+++ b/bpf/lib/proxy.h
@@ -192,7 +192,7 @@ __ctx_redirect_to_proxy(struct __ctx_buff *ctx, void *tuple __maybe_unused,
 			__be16 proxy_port, bool from_host __maybe_unused,
 			bool ipv4 __maybe_unused)
 {
-	int result = CTX_ACT_OK;
+	int result __maybe_unused = CTX_ACT_OK;
 
 #ifdef ENABLE_TPROXY
 	if (!from_host)

